contracts/citycoin-logic-v1.clar

```diff
@@ -1,967 +1,80 @@
-    )
-    (begin
-        (try! (can-claim-tokens tx-sender mined-stacks-block-ht random-sample block block-height))
-        (try! (set-tokens-claimed mined-stacks-block-ht))
-        (try! (mint-coinbase tx-sender mined-stacks-block-ht))
-
-        (ok true)
-    ))
-)
-
-;; Claim a Stacking reward.  Once a reward cycle passes, a Stacker can call this method to obtain any
-;; uSTX that were committed to the contract during that reward cycle (proportional to how many tokens
-;; they locked up).
-(define-public (claim-stacking-reward (target-reward-cycle uint))
-    (let (
-        (stacker tx-sender)
-        (cur-reward-cycle (unwrap! (get-reward-cycle block-height) (err ERR-STACKING-NOT-AVAILABLE)))
-        (entitled-ustx (get-stacking-reward tx-sender target-reward-cycle))
-        (to-return (get to-return (default-to { amount-token: u0, to-return: u0 } (get-stacked-per-cycle stacker target-reward-cycle))))
-    )
-    (begin
-        ;; check that target reward cycle is complete
-        (asserts! (> cur-reward-cycle target-reward-cycle)
-            (err ERR-CYCLE-NOT-COMPLETED))
-
-        (asserts! (or (> to-return u0) (> entitled-ustx u0))
-            (err ERR-NOTHING-TO-REDEEM))
-
-        ;; disable ability to claim again
-        (map-set stacked-per-cycle
-            { stacker: tx-sender, reward-cycle: target-reward-cycle }
-            { amount-token: u0, to-return: u0 }
-        )
-
-        ;; send back stacked tokens if user was eligible
-        (if (> to-return u0)
-            (try! (as-contract (contract-call? .citycoin-token transfer to-return tx-sender stacker none)))
-            true
-        )
-
-        ;; send back rewards if user was eligible
-        (if (> entitled-ustx u0)
-            (try! (as-contract (stx-transfer? entitled-ustx tx-sender stacker)))
-            true
-        )
-
-        (ok true)
-    ))
-)
-
-(define-read-only (get-total-supply-ustx)
-    (ok (stx-get-balance (as-contract tx-sender)))
-)
-
-(define-read-only (get-city-wallet)
-    (var-get city-wallet)
-)
-
-;; Update the city-wallet variable
-;; This can only be called by the city-wallet principal.
-;; Calling `set-city-wallet` from an outside contract (that is not the city-wallet principal)
-;; is not allowed
-(define-public (set-city-wallet (wallet-address principal))
-  (begin
-    (asserts! (is-eq contract-caller (var-get city-wallet)) (err ERR-UNAUTHORIZED))
-    (ok (var-set city-wallet wallet-address))
+      (toStackers (- amountUstx toCity))
+    )
+    (asserts! (get-activation-status) (err ERR_CONTRACT_NOT_ACTIVATED))
+    (asserts! (not (has-mined-at-block stacksHeight userId)) (err ERR_USER_ALREADY_MINED))
+    (asserts! (> amountUstx u0) (err ERR_INSUFFICIENT_COMMITMENT))
+    (asserts! (>= (stx-get-balance tx-sender) amountUstx) (err ERR_INSUFFICIENT_BALANCE))
+    (unwrap-panic (contract-call? .citycoin-core set-tokens-mined userId stacksHeight amountUstx toStackers toCity))
```

Comment:
I think we should avoid `unwrap-panic` whenever it is possible.
Here if something goes wrong while executing `set-tokens-mined`, TX will fail with runtime error and we'll have to manually investigate what happen. 
If we wrap it with `(try! )` all errors thrown in `set-tokens-mined` will be propagated here and as a result TX will fail with meaningful error.

Reply:
Good call! I remember we ran into this in earlier stages of development. Will swap to `try!`.